README.md
# @mconnect/datapro

- Data processing operations - analytics, stat, machine-learning etc.
- Includes data analytics, ETL, statistics and machine-learning functions
- Document all functions / features, with detailed test-cases (i.e. TDD)

## Installation

```sh

$ nimble install mcdata

```

## API

- Data analysis and transformation

```js
// Coming soon

const abc = 'test';
console.log('abc: ', abc);

```

## Examples

```js
// Coming soon

const abc = 'test';
console.log('abc: ', abc);

```


HISTORY.md
# Historical logs

## Version 0.1.0

- Release date: May 30, 2020
- includes data analytics, ETL, statistics and machine-learning functions
- statistics and machine learning functions are work-in-progress


tests/testcollections.nim
# This is just an example to get you started. You may wish to put all of your
# tests into a single file, or separate them into multiple `test1`, `test2`
# etc. files (better names are recommended, just make sure the name starts with
# the letter 't').
#
# To run these tests, simply execute `nimble test`.

import unittest

import mcdata

echo "Collections (Array) Filtering"
    
const 
    myArray = @[10, 20, 31, 40, 51, 63, 70, 80, 95, 105, 110, 121, 150]
    myArray2 = @["Abiodun", "Olawale", "Abi", "Ola", "Akindele", "Olayemi", "Vic"]
    
# ***Implementation - examples***

# IPredicate implementation
proc myFilterFunc(val: int): bool =
  val mod 2 == 0

# StringPredicate implementation
proc myFilterFunc(val: string): bool =
  len(val) > 5

proc myFilterStringFunc(val: string): bool =
  len(val) > 5

test "Filter: specific and generic functions: ":
  check $(filterFunc(myArray, myFilterFunc)) == "@[10, 20, 40, 70, 80, 110, 150]"
  check filterFunc(myArray, myFilterFunc) == @[10, 20, 40, 70, 80, 110, 150]
  check filterFunc(myArray2, myFilterStringFunc) == @["Abiodun", "Olawale", "Akindele", "Olayemi"]
  check filterFunc[int, IPredicate](myArray, myFilterFunc) == @[10, 20, 40, 70, 80, 110, 150]
  check filterFunc[string, StringPredicate](myArray2, myFilterStringFunc) == @["Abiodun", "Olawale", "Akindele", "Olayemi"]
  check filterFunc[int, Predicate[int]](myArray, myFilterFunc) == @[10, 20, 40, 70, 80, 110, 150]
  check filterFunc[string, Predicate[string]](myArray2, myFilterStringFunc) == @["Abiodun", "Olawale", "Akindele", "Olayemi"]


test "generic filter and specific action functions: ":
  check filterFunc[int, IPredicate](myArray, myFilterFunc) == @[10, 20, 40, 70, 80, 110, 150]
  check filterFunc[string, StringPredicate](myArray2, myFilterStringFunc) == @["Abiodun", "Olawale", "Akindele", "Olayemi"]
  
test "generic filter and action functions: ":
  check filterFunc[int, Predicate[int]](myArray, myFilterFunc) == @[10, 20, 40, 70, 80, 110, 150]
  check filterFunc[string, Predicate[string]](myArray2, myFilterStringFunc) == @["Abiodun", "Olawale", "Akindele", "Olayemi"]

# test "can add":
#   check add(5, 5) == 10

src/mcdata/collections.nim
#
#            mconnect collections package
#        (c) Copyright 2020 Abi Akindele (mconnect.biz)
#
#    See the file "LICENSE.md", included in this
#    distribution, for details about the copyright / license.
#

import sequtils, sugar

##[The collections interface/types for handling various collection tasks:
  Predicate, Function, Operation, Consumer, Supplier etc.
  ]##
type
    IPredicate* = proc(val: int): bool
    StringPredicate* = proc(val: string): bool
    Predicate*[T] = proc(val: T): bool
    BinaryPredicate*[T, U] = proc(val1: T, val2: U ): bool
    UnaryOperator*[T] = proc(val: T): T
    BinaryOperator*[T] = proc(val1, val2: T): T
    Function*[T, R] = proc(val: T): R
    BiFunction*[T, U, R] = proc(val1: T, val2: U): R
    Consumer*[T] = proc(val: T)
    BiConsumer*[T, U] = proc(val1: T, val2: U)
    Supplier*[R] = proc(): R
    Comparator*[T] = proc(val1: T, val2: T): int

# Collection procedures/functions - overloaded procedures

# Generic filter function - using Predicate type
proc filterFunc*[T, U](arr: seq[T], filterFn: U): seq[T] =
    # functional style
    # result = arr.filter((x:T)->bool => filterFn(x))
    result = arr.filter(it => filterFn(it))

# Integer filter function - using IPredicate type
proc filterFunc*(arr: seq[int], filterFn: IPredicate): seq[int] =
    var myResult: seq[int] = @[]
    
    for i in arr:
        if(filterFn(i)):
            myResult.add(i)
    result = myResult

# String filter function - using StringPredicate type
proc filterFunc*(arr: seq[string], filterFn: StringPredicate): seq[string] =
        var myResult: seq[string] = @[]
        # echo "arr-result, start: ", myResult
        
        for i in arr:
            if(filterFn(i)):
                myResult.add(i)
        result = myResult
        # echo "arr-result, final: ", myResult    

# TODO: define other collection types: BinaryPredicate, UnaryOperator, BinaryOperator,
# TODO: Function, BiFunction, Consumer, BiConsumer, Supplier, Comparator...

# Run the following test, if mainModule
when isMainModule:
    # ***Implementation - examples***
    const 
        myArray = @[10, 20, 31, 40, 51, 63, 70, 80, 95, 105, 110, 121, 150]
        myArray2 = @["Abiodun", "Olawale", "Abi", "Ola", "Akindele", "Olayemi", "Vic"]
  
    # IPredicate implementation
    proc myFilterFunc(val: int): bool =
            val mod 2 == 0

    # StringPredicate implementation
    proc myFilterStringFunc(val: string): bool =
        len(val) > 5

    echo "specific type filter and action functions"
    echo "******************************"
    echo filterFunc(myArray, myFilterFunc)
    echo $(filterFunc(myArray, myFilterFunc)) == "@[10, 20, 40, 70, 80, 110, 150]"
    echo filterFunc(myArray, myFilterFunc) == @[10, 20, 40, 70, 80, 110, 150]
    doAssert filterFunc(myArray, myFilterFunc) == @[10, 20, 40, 70, 80, 110, 150]
    echo filterFunc(myArray2, myFilterStringFunc)
    echo filterFunc(myArray2, myFilterStringFunc) == @["Abiodun", "Olawale", "Akindele", "Olayemi"]
    # echo $(filterFunc(myArray2, myFilterStringFunc) == "@['Abiodun', 'Olawale', 'Akindele', 'Olayemi']"
    doAssert filterFunc(myArray2, myFilterStringFunc) == @["Abiodun", "Olawale", "Akindele", "Olayemi"]
    echo ""
    echo "generic filter and specific action functions"
    echo "******************************"
    echo filterFunc[int, IPredicate](myArray, myFilterFunc)
    echo filterFunc[string, StringPredicate](myArray2, myFilterStringFunc)
    echo ""
    echo "generic filter and action functions"
    echo "******************************"
    echo filterFunc[int, Predicate[int]](myArray, myFilterFunc)
    echo filterFunc[string, Predicate[string]](myArray2, myFilterStringFunc)
    echo ""
